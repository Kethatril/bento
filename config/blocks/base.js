module.exports = function(config, options, { select }) {
    config.merge({
        context: options.get("paths.input"),
        entry: options.get("entry"),

        devtool: select({
            development: "cheap-module-eval-source-map",
            production: options.get("enabled.sourceMap")
                ? "source-map"
                : undefined
        }),

        output: {
            path: options.get("paths.output"),
            publicPath: options.get("paths.public"),
            filename: `scripts/${options
                .get("filename")
                .replace("hash", "chunkhash")}.js`
        },

        resolve: options.get("resolve")
    });
};
