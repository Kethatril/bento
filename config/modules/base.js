const { config, select, read } = require("../utils");

module.exports = config(instance => {
    return instance.merge({
        context: read("paths.resolved.input"),
        entry: read("entry"),

        devtool: select({
            development: "cheap-module-eval-source-map",
            production: read("enabled.sourceMap") ? "source-map" : undefined
        }),

        output: {
            path: read("paths.resolved.output"),
            publicPath: read("paths.public"),
            filename: `scripts/${read("filename").replace(
                "hash",
                "chunkhash"
            )}.js`
        },

        resolve: read("resolve")
    });
});
