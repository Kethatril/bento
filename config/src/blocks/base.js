import merge from "webpack-merge";

export default function(config, options, { select }) {
    return merge(config, {
        context: options.get("paths.input"),
        entry: options.get("entry"),

        devtool: select({
            development: "cheap-module-eval-source-map",
            production: options.get("enabled.sourceMap")
                ? "source-map"
                : undefined
        }),

        output: {
            path: options.get("paths.output"),
            publicPath: options.get("paths.public"),
            filename: `scripts/${options
                .get("filename")
                .replace("hash", "chunkhash")}.js`
        },

        resolve: options.get("resolve")
    });
}
