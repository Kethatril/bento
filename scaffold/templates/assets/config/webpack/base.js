module.exports = {
    /**
     * ------------------------------------------------------------------------
     * Environment
     * ------------------------------------------------------------------------
     * "value":     The environment value that should be used to determine
     *              various aspects of webpack config, eg `devtool` value
     *
     */
    env: {
        value: process.env.NODE_ENV || "development",
        isProduction: process.env.NODE_ENV === "production",
        isDevServer: process.argv[1].indexOf("webpack-dev-server") >= 0
    },

    /**
     * ------------------------------------------------------------------------
     * Paths
     * ------------------------------------------------------------------------
     * "root":      The base directory, an absolute path, that will be used
     *              to resolved all other relative paths in here.
     *
     * "input":     The base directory for webpack to looking for resolving
     *              entry points and loaders from configuration. It's relative
     *              to `paths.root` [1]
     *
     * "output":    The output directory. It's relative to `paths.root` [2]
     *
     * "public":    The public URL of the output directory when referenced
     *              in a browser. [3]
     *
     * [1] @see https://webpack.js.org/configuration/entry-context/#context
     * [2] @see https://webpack.js.org/configuration/output/#output-path
     * [3] @see https://webpack.js.org/configuration/output/#output-publicpath
     *
     */
    paths: {
        root: process.cwd(),
        input: "<%= input %>",
        output: "<%= output %>",
        public: "<%= public %>"
    },

    /**
     * ------------------------------------------------------------------------
     * Entry
     * ------------------------------------------------------------------------
     * "main":      The point or points to enter the application.
     *              Note: vendor static css files can be added here instead of
     *              using standard `import` to improve HMR reload speed. [1]
     *
     * [1] @see https://webpack.js.org/configuration/entry-context/#context
     *
     */
    entry: {
        main: [
            // "normalize.css/normalize.css"
            "./scripts/main.js"
        ]
    },

    /**
     * ------------------------------------------------------------------------
     * Resolve
     * ------------------------------------------------------------------------
     * Webpack's `resolve` options can be defined here. [1]
     *
     * By default "~" is resolved to `scripts` folder, make it easier to
     * import modules within there. [2]
     *
     * [1] @see https://webpack.js.org/configuration/resolve/
     * [2] @see https://webpack.js.org/configuration/resolve/#resolve-alias
     *
     */
    resolve: {
        alias: {
            "@": "@{paths.input}",
            "~": "@{paths.input}/scripts"
        }
    },

    /**
     * ------------------------------------------------------------------------
     * Caching
     * ------------------------------------------------------------------------
     * "hash":      Hash length that will append to output files for
     *              long term caching. [1]
     *
     * [1] @see https://webpack.js.org/guides/caching/#the-problem
     *
     */
    caching: {
        hash: ".[hash:8]"
    },

    /**
     * ------------------------------------------------------------------------
     * Additional options
     * ------------------------------------------------------------------------
     * sourceMap: Enable / disable source map
     *
     */
    enabled: {
        sourceMap: "@{env.isDevServer}",
        extractCSS: "@{env.isProduction}",
        writeManifest: true
    },

    /**
     * ------------------------------------------------------------------------
     * Files
     * ------------------------------------------------------------------------
     * "copy":      Copy files that match the pattern to the output folder.
     *              Note: glob path is relative to `paths.input`
     *
     */
    files: {
        copy: "+(images|media)/**/*"
    }
};
