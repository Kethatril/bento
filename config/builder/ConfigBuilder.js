const { Config } = require("webpack-config"),
    parse = require("./modules/parse");

module.exports = class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    constructor(options = {}) {
        this._options = parse(options);
        this._config = new Config();
        this._modules = [];
        this._manifest = {};
    }

    use(...modules) {
        this._modules = this._modules.concat(modules);
        return this;
    }

    build(fn) {
        this._modules.push(fn);

        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._modules,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            let config = new Config(),
                options = this._options;

            try {
                (function next(err) {
                    let async;

                    // Increment counter variable and skip any indices that
                    // don't exist. This allows sparse arrays to be iterated.
                    do {
                        ++i;
                    } while (!(i in modules) && i !== len);

                    if (err === false || err) return reject(err);

                    if (i === len) return resolve(config.toObject());

                    let result;

                    result = modules[i].call(
                        {
                            async() {
                                async = true;
                                return next;
                            }
                        },
                        config,
                        options
                    );
                    if (!async) {
                        next(result);
                    }
                })();
            } catch (e) {
                reject(e);
            }
        });
    }
};
