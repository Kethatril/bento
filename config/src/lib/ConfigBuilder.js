import { parse, resolve, select } from "~/lib/utils";
import ManifestKeeper from "~/lib/ManifestKeeper";

export default class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    manifest = new ManifestKeeper();

    constructor(options = {}) {
        this._options = parse(resolve(options));
        this._blocks = options.blocks.list || [];
        this._timeout = options.blocks.timeout || 3000;
    }

    get options() {
        return this._options;
    }

    build() {
        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._blocks,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            const timeout = this._timeout;

            let config = {},
                options = this._options,
                utils = { select: select(options), manifest: this.manifest };

            (function next(result) {
                let async;

                // Increment counter variable and skip any indices that
                // don't exist. This allows sparse arrays to be iterated.
                do {
                    ++i;
                } while (!(i in modules) && i !== len);

                if (result === false || result instanceof Error)
                    return reject(err);

                if (i === len) {
                    return resolve(result);
                }

                try {
                    result = modules[i](result, options, utils);
                } catch (e) {
                    return reject(e);
                }
                if (!result || typeof result !== "object") {
                    throw new Error(
                        "[InvalidValueError] Each block must either return a new config object or a promise that resolves a new config object."
                    );
                }

                if (result.then) {
                    const id = setTimeout(function() {
                        if (!async) return;
                        return reject(
                            new Error(
                                `[TimeoutError] ${timeout} ms exceeded. For async blocks, ensure the returned promise is resolved.`
                            )
                        );
                    }, timeout);

                    const done = function(result) {
                        clearTimeout(id);
                        next(result);
                    };

                    async = true;
                    result.then(done, done);
                }

                if (!async) {
                    next(result);
                }
            })(config);
        });
    }
}
