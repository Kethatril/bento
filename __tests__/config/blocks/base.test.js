import base from "~/blocks/base";
import ConfigBuilder from "~/lib/ConfigBuilder";

describe("blocks/base", () => {
    const options = {
        paths: {
            root: process.cwd(),
            input: "./app",
            output: "./public",
            public: "/"
        },

        entry: {
            main: ["./main.js"]
        },

        filename: "[name]",

        blocks: {
            list: [base]
        },

        resolve: {}
    };

    it("matches snapshot when in development", () => {
        const builder = ConfigBuilder.create(
            Object.assign(
                {},
                {
                    env: {
                        value: "development"
                    }
                },
                options
            )
        );

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });

    it("matches snapshot when in production", () => {
        const builder = ConfigBuilder.create(
            Object.assign(
                {},
                {
                    env: {
                        value: "production"
                    }
                },
                options
            )
        );

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });
});
