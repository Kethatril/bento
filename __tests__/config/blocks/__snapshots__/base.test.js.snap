// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks/base matches snapshot when in development 1`] = `
Object {
  "context": "<cwd>/app",
  "devtool": "cheap-module-eval-source-map",
  "entry": Object {
    "main": Array [
      "./main.js",
    ],
  },
  "output": Object {
    "filename": "scripts/[name].js",
    "path": "<cwd>/public",
    "publicPath": "/",
  },
  "resolve": Object {},
}
`;

exports[`blocks/base matches snapshot when in production 1`] = `
Object {
  "context": "<cwd>/app",
  "devtool": false,
  "entry": Object {
    "main": Array [
      "./main.js",
    ],
  },
  "output": Object {
    "filename": "scripts/[name].js",
    "path": "<cwd>/public",
    "publicPath": "/",
  },
  "resolve": Object {},
}
`;
